isocroft-syntax.md

```diff
@@ -0,0 +1,913 @@
+    }
+    
+    func get_number(int num) -> int {
+        return 20
+    }
+    ```
+
+- An example of a type identifier. More on types [later](#types).
+
+    ```swift
+    type Person {
+        name: str,
+        age: int,
+    }
+    ```
+    
+## ARITHMETIC OPERATION
+- Operations in ratio can either be arithmetic, bitwise, logical, relational or functional
+
+####  functional opearator (binary)
+
+* `~>` (pipe)
+
+#### aritmetic operators (binary)
+
+* `+` (add)
+* `-` (subtract)
+* `*` (multiply)
+* `/` (divide)
+* `%` (modulo)
+* `+=` (rewrite add)
+* `-=` (rewrite subtract)
+
+#### arithmetic operators (unary)
+
+* `++` (increment)
+* `--` (decrement)
+
+#### logical operators (binary)
+
+* `&&` (and)
+* `and` (and)
+* `||` (or)
+* `or` (or)
+
+#### logical operators (unary)
+
+* `!` (not)
+* `not` (not)
+
+#### relational operators (binary)
+
+* `<`
+* `>`
+* `>=`
+* `<=`
+* `!=`
+* `!==`
+
+#### bitwise operators (binary)
+
+* `>>`
+
+```js
+
+var item = 3
+
+print(--item) // 2
+print(++item) // 3
+
+```
+
+## ARRAYS ( Mutable / Immutable )
+- Arrays in ratio will serve as an array of variables of unmixed or mixed types
+
+>mutable arrays (lists) are defined using `var`
+```js
+
+var int<4> list = [1, 2, 3, 4]
+
+list[0] = 0 // No error
+
+print(list); // 0, 2, 3, 4
+
+```
+
+>immutable arrays (tuples) are defined using `const`
+
+```js
+
+const int<2> tuple = [1, 5]; # you can't add or remove from the tuple but you can change the content at all indexes
+
+tuple[0] = 2 // No error
+
+print(tuple); // 2, 5
+```
+
+- Array of a complex type in ratio
+
+```js
+
+var Person<3> persons = [ 
+   {name: "Tochukwu Nkedilim", age: 27}, 
+   {name: "Ikechi Micheal", age: 29}, 
+   {name: "Azeez Adewale", age: 26} 
+]
+```
+
+## SEMI COLONS
+- Semi colons are used to separate multiple statements or expressions.
+
+    ```swift
+    const name = "John"; const age = 40;
+    print(name, age);
+    ```
+
+- Semi colons are optional if statements/expressions are separated by newlines.
+
+    ```cpp
+    const name = "John"
+    const age = 40
+    print(name, age)
+    ```
+
+## TYPE ANNOTATION
+- Types can be automatically inferred in Ratio.
+    ```js
+    var todo = "Create a new programming language :)" // `todo` is inferred as `str` (String)
+    ```
+
+    ```cpp
+    func add(a, b) {
+        return a + b
+    }
+
+    const sum = add(1, 1) // Usage of `add` inferred as `(int, int) -> int` here.
+    ```
+
+- Subjects can be explicitly type-annotated.
+
+    ```js
+    var int identification = 60
+    ```
+
+- Functions can also be explicitly type-annotated.
+
+    ```swift
+    func add(a int, b int) -> int {
+        return a + b
+    }
+    ```
+
+## CONCATENATION ( OVERLOADED OPERATOR - `++` )
+
+- The concatenation symbol for ratio is `++`
+
+```swift
+func sort(arr) -> int {
+   
+}
+```
+
+>the pipe operator `~>` can also be used to sort the resulting list
+```elixir
+
+var int<8> numbers = [1,3,5,7] ++ [0,2,4,6] ~> sort
```

Comment:
You are correct... At compile time, this program will raise an error in a language like Java or C++ as at compile time the size of the array cannot be determined until runtime. 


But i wrote that line in the hope that we could design a multipass and AST-driven compiler such that the AST can be modified at the entry point into the *Code Generation* (Backend) to find expressions that can be "executed" or "reduced statically" to a literal value (So in the example above, the AST nodes are reduced to "[" "1" "," "3" "," "5" "," "7" "," "0" "," "2" "," "4" "," "6" "]" "~" ">" "sort"). 


However this feature will work for only literal primitives and not non-literal primitives to avoid unbearable/excessive overhead to discover such expressions and "reduce" them. There are other use cases for such a feature... but this is just a simple example.